internal/gps/source_cache_test.go

```diff
@@ -0,0 +1,426 @@
+		if !ok {
+			t.Errorf("no package tree found:\n\t(WNT): %#v", pt)
+		}
+		comparePackageTree(t, pt, got)
+	})
+
+	t.Run("versions", func(t *testing.T) {
+		c := newMemoryCache()
+
+		const rev1, rev2 = "rev1", "rev2"
+		const br, ver = "branch_name", "2.10"
+		versions := []PairedVersion{
+			NewBranch(br).Pair(rev1),
+			NewVersion(ver).Pair(rev2),
+		}
+		SortPairedForDowngrade(versions)
+		c.storeVersionMap(versions, true)
+
+		t.Run("getAllVersions", func(t *testing.T) {
+			got := c.getAllVersions()
+			if len(got) != len(versions) {
+				t.Errorf("unexpected versions:\n\t(GOT): %#v\n\t(WNT): %#v", got, versions)
+			} else {
+				SortPairedForDowngrade(got)
+				for i := range versions {
+					if !versions[i].equals(got[i]) {
+						t.Errorf("unexpected versions:\n\t(GOT): %#v\n\t(WNT): %#v", got, versions)
+						break
+					}
+				}
+			}
+		})
+
+		revToUV := map[Revision]UnpairedVersion{
+			rev1: NewBranch(br),
+			rev2: NewVersion(ver),
+		}
+
+		t.Run("getVersionsFor", func(t *testing.T) {
+			for rev, want := range revToUV {
+				rev, want := rev, want
+				t.Run(string(rev), func(t *testing.T) {
+					uvs, ok := c.getVersionsFor(rev)
+					if !ok {
+						t.Errorf("no version found:\n\t(WNT) %#v", want)
+					} else if len(uvs) != 1 {
+						t.Errorf("expected one result but got %d", len(uvs))
+					} else {
+						uv := uvs[0]
+						if uv.Type() != want.Type() {
+							t.Errorf("expected version type %d but got %d", want.Type(), uv.Type())
+						}
+						if uv.String() != want.String() {
+							t.Errorf("expected version %q but got %q", want.String(), uv.String())
+						}
+					}
+				})
+			}
+		})
+
+		t.Run("getRevisionFor", func(t *testing.T) {
+			for want, uv := range revToUV {
+				want, uv := want, uv
+				t.Run(uv.String(), func(t *testing.T) {
+					rev, ok := c.getRevisionFor(uv)
+					if !ok {
+						t.Errorf("expected revision %q but got none", want)
+					} else if rev != want {
+						t.Errorf("expected revision %q but got %q", want, rev)
+					}
+				})
+			}
+		})
+
+		t.Run("toRevision", func(t *testing.T) {
+			for want, uv := range revToUV {
+				want, uv := want, uv
+				t.Run(uv.String(), func(t *testing.T) {
+					rev, ok := c.toRevision(uv)
+					if !ok {
+						t.Errorf("expected revision %q but got none", want)
+					} else if rev != want {
+						t.Errorf("expected revision %q but got %q", want, rev)
+					}
+				})
+			}
+		})
+
+		t.Run("toUnpaired", func(t *testing.T) {
+			for rev, want := range revToUV {
+				rev, want := rev, want
+				t.Run(want.String(), func(t *testing.T) {
+					uv, ok := c.toUnpaired(rev)
+					if !ok {
+						t.Errorf("no UnpairedVersion found:\n\t(WNT): %#v", uv)
+					} else if !uv.equals(want) {
+						t.Errorf("unexpected UnpairedVersion:\n\t(GOT): %#v\n\t(WNT): %#v", uv, want)
+					}
+				})
+			}
+		})
+	})
+}
+
+// compareManifests compares two manifests and reports differences as test errors.
+func compareManifests(t *testing.T, want, got Manifest) {
+	{
```

Comment:
Yeah, it seemed cleaner than a bunch of want- _Type_ vars.

In general, I've found it handy to isolate small portions of longer functions that aren't candidates for entirely separate functions, if only to help readability regarding variable scope or to give some better scope to a comment than line breaks.